python/pylibvw.cc

```diff
@@ -56,6 +57,117 @@ const size_t pACTION_PDF_VALUE = 9;
 
 void dont_delete_me(void* arg) {}
 
+class OptionManager
+{
+  std::map<std::string, std::vector<VW::config::option_group_definition>> m_option_group_dic;
+  py::object m_py_opt_class;
+  VW::config::options_i& m_opt;
+  std::vector<std::string>& m_enabled_reductions;
+  std::string default_group_name;
+
+public:
+  OptionManager(VW::config::options_i& options, std::vector<std::string>& enabled_reductions, py::object py_class)
+      : m_opt(options)
+      , m_enabled_reductions(enabled_reductions)
+      , m_option_group_dic(options.get_collection_of_options())
+      , m_py_opt_class(py_class)
+  {
+    default_group_name = static_cast<VW::config::options_boost_po*>(&options)->m_default_tint;
+  }
+
+  template <typename T>
+  py::object* value_to_pyobject(VW::config::typed_option<T>& opt)
+  {
+    return new py::object(
+        m_py_opt_class(opt.m_name, opt.m_help, opt.m_short_name, opt.m_keep, opt.m_necessary, opt.m_allow_override,
+            opt.value(), m_opt.was_supplied(opt.m_name), opt.default_value(), opt.default_value_supplied()));
+  }
+
+  template <typename T>
+  py::object* value_to_pyobject(VW::config::typed_option<std::vector<T>>& opt)
+  {
+    py::list values;
+
+    auto vec = opt.value();
+    if (vec.size() > 0)
+    {
+      for (auto const& opt : vec) { values.append(py::object(opt)); }
+    }
+
+    return new py::object(m_py_opt_class(opt.m_name, opt.m_help, opt.m_short_name, opt.m_keep, opt.m_necessary,
+        opt.m_allow_override, values, m_opt.was_supplied(opt.m_name), py::list(), opt.default_value_supplied()));
+  }
+
+  template <typename T>
+  py::object* transform_if_t(VW::config::base_option& base_option)
+  {
+    if (base_option.m_type_hash == typeid(T).hash_code())
+    {
+      auto typed = dynamic_cast<VW::config::typed_option<T>&>(base_option);
+      return value_to_pyobject(typed);
+    }
+
+    return nullptr;
+  }
+
+  template <typename TTypes>
+  py::object base_option_to_pyobject(VW::config::base_option& options)
+  {
+    py::object* temp = transform_if_t<typename TTypes::head>(options);
+    if (temp != nullptr)
+    {
+      auto repack = py::object(*temp);
```

Comment:
why does this need to be repacked?

Reply:
not a fan of this chunk, but it is to unify all the `py::objects` on the same memory mgmt scheme. If you see `line 97`, this object is being allocated with `new` expression, which would imply that someone is responsible for destroying it later. Boost python api provides `manage_new_object` for these cases, but it gets a bit complicated since I'm also using a `py::dict` not using new down in `line 129`. So my workaround was to instantiate here another copy and deleting the original. Let me know if you have more thoughts about this.

Reply:
I see. Is there a way to avoid the `new` in line 97 or do you have to create the object using `new`?

Reply:
I can't think of another way of doing it